### ISSUE TITLE
Concurrent modification exception when querying for Azure Managed Disks.

### ISSUE THREAD

**Issue by ImranBijapuri** [Azure/azure-sdk-for-java#1942]
I am using Azure Java SDK 1.1.2. But while querying for managed disks I get the following error stack trace. According to Microsoft this is because of using Java 8 while the SDK was developed using Java 7 features (https://github.com/Azure/azure-sdk-for-java/issues/1877). I still don't think it could be a Java 8 issue.

The error occurs from this line of code

`Azure().disks().list().stream().collect(Collectors.toSet()`

```
2017-10-05 03:27:17,516 ERROR [sdk-embed-op-64] [AzureProbe]    : Exception in discovery Microsoft Azure Enterprise
java.util.ConcurrentModificationException
    at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:901) ~[?:1.8.0_102]
    at java.util.ArrayList$Itr.next(ArrayList.java:851) ~[?:1.8.0_102]
    at com.microsoft.azure.PagedList$ListItr.next(PagedList.java:170) ~[azure-client-runtime-1.1.1-SNAPSHOT.jar:?]
    at java.util.Iterator.forEachRemaining(Iterator.java:116) ~[?:1.8.0_102]
    at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801) ~[?:1.8.0_102]
    at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_102]
    at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_102]
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[?:1.8.0_102]
    at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_102]
    at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[?:1.8.0_102]
```

**Comment by anuchandy** [comment]
@ImranBijapuri 

I think after some investigation I agree with you - the concurrent modification has nothing to do with the java8.

Here is a workaround to unblock you:

```java
PagedList<Disk> list = Azure().disks().list();
list.loadAll();
list.stream().collect(Collectors.toSet());
```

To idea is invoke `loadAll()` on the list before using the java8 stream created from the list.

I was initially bit concerned about the performance due to the invocation of `loadAll` (which loads all the pages by making one API call per page) but it seems any built-in operation you use with the stream backed by a list invokes `size()` method on the list iterator. The `size` implementation on the paged list iterator invokes `loadAll` anyway. 

I feel like, java8 stream is not optimized for lazy list (list where the size is not known upfront), because a simple iteration on the stream using built-in forEach method requires the knowledge of backing list size.

```java
        list.stream().forEach(new Consumer<String>() {
            @Override
            public void accept(String item) {
                System.out.println(item);
            }
        });
```

The above iteration invokes list `size()` method.

If you are looking for asynchronous lazy page, i would recommend you to use `Observable<Disk> listAsync()` instead.

**Comment by tsarna** [comment]
TL;DR doing loadAll is an OK workaround, but it's bug in PagedList that one can't create an iterator and then take the size of the list.

---

Here's why it doesn't work without the loadAll():

The returned list is a com.microsoft.azure.PagedList. PagedList does implement any of the new Java 8 list methods of course, so you get a series of default implementations from the declared interfaces (stream() from Collection, spliterator() from List, etc). Internally this ends up creating an instance of (the wonderfully named) IteratorSpliterator from java.util.Spliterators on the PagedList, and calling forEachRemaining() on it, and here is where the problem lies.

IteratorSpliterator's forEachRemaining creates an iterator on the PagedList, and then gets the paged lists's size(). It could just as easily do these in the opposite order and then everything would be fine, but it doesn't. On the other hand, there's no reason this order shouldn't work with any List implementation, because size() on a list should not invalidate an iterator on that list.

However, it does for PagedList. The iterator on the PagedList wraps an iterator on the PagedList's internal wrapped list. The size() call on the pageList does loadAll(), which modifies the wrapped list. Then, it tries to use the iterator it has created, which uses the wrapped iterator on the now-modified wrapped list, which throws the ConcurrentModificationException.

The PagedList's iterator has some code to manage creating a new wrapped iterator when it loads a new page, so this doesn't cause a problem in the most trivial uses. But, it's not really a complete solution as it falls down in slightly more complicated scenarios, such as calling size() as here, or for example creating multiple iterators on the same PagedList.

**Comment by anuchandy** [comment]
@tsarna Thank you for investigating this!

Upon calling size() refreshing all the iterator created from the `PagedList` requires parent list to keep track of all the iterator created from it. We need to think more on this. For now, we are postponing further work on `PagedList` because of the complexity involved, may be we will consider it in the next major version upgrade or as part of switching to java8, but no ETA. 

Mean time, if you have some idea to improve this data structure to support these scenarios and want to contribute, feel free to open a PR to our runtime https://github.com/Azure/autorest-clientruntime-for-java.

**Comment by anuchandy** [comment]
Submitted a fix https://github.com/Azure/autorest-clientruntime-for-java/pull/258

**Comment by anuchandy** [comment]
Runtime with fix has been released v1.1.1 https://github.com/Azure/autorest-clientruntime-for-java/tree/v1.1.1 hence closing this issue.
